core/src/main/java/tachyon/worker/rdma/RDMADataServer.java

```diff
@@ -0,0 +1,236 @@
+  }
+
+  /**
+   * Callbacks for the listener server portal
+   */
+  private final class PortalServerCallbacks implements ServerPortal.Callbacks {
+
+    public void onSessionEvent(EventName session_event, EventReason reason) {
+      LOG.debug("got event {} because of {}", session_event.toString(), reason.toString());
+      if (session_event == EventName.PORTAL_CLOSED) {
+        mEventQueueHandler.breakEventLoop();
+      }
+    }
+
+    public void onSessionNew(SessionKey sesKey, String srcIP, Worker workerHint) {
+      LOG.info("onSessionNew {}", sesKey.getUri());
+      URI uri;
+      try {
+        uri = new URI(sesKey.getUri());
+      } catch (URISyntaxException  e) {
+        mListener.reject(sesKey, EventReason.ADDR_ERROR, "Client uri could not be resolved");
+        return;
+      }
+      SessionServerCallbacks callbacks = new SessionServerCallbacks(uri);
+      ServerSession session = new ServerSession(sesKey, callbacks);
+      callbacks.setSession(session);
+      mListener.accept(session);
+    }
+  }
+
+  private final class SessionServerCallbacks implements ServerSession.Callbacks {
+    private final DataServerMessage mResponseMessage;
+    private ServerSession mSession;
+
+    public SessionServerCallbacks(URI uri) {
+      List<NameValuePair> params = URLEncodedUtils.parse(uri, "UTF-8");
+      long blockId = Long.parseLong(params.get(0).getValue());
+      long offset = Long.parseLong(params.get(1).getValue());
+      long length = Long.parseLong(params.get(2).getValue());
+      LOG.debug("got request for block id {} with offset {} and length {}",
+          blockId, offset, length);
+      final int lockId = mBlocksLocker.getLockId();
+      final StorageDir storageDir = mBlocksLocker.lock(blockId, lockId);
+      ByteBuffer data;
+      int dataLen = 0;
+      try {
+        data = storageDir.getBlockData(blockId, offset, (int)length);
+        storageDir.accessBlock(blockId);
+        dataLen = data.limit();
+      } catch (Exception e) {
+        LOG.error(e.getMessage(), e);
+        data = null;
+      }
+      mResponseMessage =
+          DataServerMessage.createBlockResponseMessage(true, blockId, offset, dataLen, data);
+      mResponseMessage.setLockId(lockId);
+    }
+
+    public void setSession(ServerSession ses) {
+      mSession = ses;
+    }
+
+    public void onRequest(Msg m) {
+      if (mSession.getIsClosing()) {
+        mSession.discardRequest(m);
+      } else {
+        mResponseMessage.copyMsgToBuffer(m.getOut());
+        try {
+          mSession.sendResponse(m);
+        } catch (JxioGeneralException e) {
+          LOG.error("Exception accured while sending messgae {}", e.toString());
+          mSession.discardRequest(m);
+        } catch (JxioSessionClosedException e) {
+          LOG.error("session was closed unexpectedly {}", e.toString());
+          mSession.discardRequest(m);
+        }
+      }
+
+      if (!mSession.getIsClosing() && mResponseMessage.finishSending()) {
```

Comment:
this block can be merged in the above `else` block since logic of `!mSession.getIsClosing()` has been contained in `else` block.